hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/persistence/ContainerSchemaManager.java

```diff
@@ -113,40 +142,121 @@ public void insertUnhealthyContainerRecords(List<UnhealthyContainers> recs) {
+    List<ContainerHistory> resList = new ArrayList<>();
+    for (Map.Entry<UUID, ContainerReplicaHistory> entry : resMap.entrySet()) {
+      final UUID uuid = entry.getKey();
+      String hostname = "N/A";
+      // Attempt to retrieve hostname from NODES table
+      if (nodeDB != null) {
+        try {
+          DatanodeDetails dnDetails = nodeDB.get(uuid);
+          if (dnDetails != null) {
+            hostname = dnDetails.getHostName();
+          }
+        } catch (IOException ex) {
+          LOG.debug("Unable to retrieve from NODES table of node {}. {}",
+              uuid, ex.getMessage());
+        }
+      }
+      final long firstSeenTime = entry.getValue().getFirstSeenTime();
+      final long lastSeenTime = entry.getValue().getLastSeenTime();
+      resList.add(new ContainerHistory(containerID, uuid.toString(), hostname,
+          firstSeenTime, lastSeenTime));
+    }
+    return resList;
   }
 
   public List<ContainerHistory> getLatestContainerHistory(long containerID,
                                                           int limit) {
-    DSLContext dslContext = containerSchemaDefinition.getDSLContext();
-    // Get container history sorted in descending order of last report timestamp
-    return dslContext.select()
-        .from(CONTAINER_HISTORY)
-        .where(CONTAINER_HISTORY.CONTAINER_ID.eq(containerID))
-        .orderBy(CONTAINER_HISTORY.LAST_REPORT_TIMESTAMP.desc())
-        .limit(limit)
-        .fetchInto(ContainerHistory.class);
+    List<ContainerHistory> res = getAllContainerHistory(containerID);
+    res.sort(comparingLong(ContainerHistory::getLastSeenTime).reversed());
+    return res.stream().limit(limit).collect(Collectors.toList());
+  }
+
+  /**
+   * Should only be called once during ReconContainerManager init.
+   */
+  public void setReplicaHistoryMap(
+      Map<Long, Map<UUID, ContainerReplicaHistory>> replicaHistoryMap) {
+    this.replicaHistoryMap = replicaHistoryMap;
+  }
+
+  /**
+   * Should only be called once during ReconContainerManager init.
+   */
+  public void setScmDBStore(DBStore scmDBStore) {
+    this.scmDBStore = scmDBStore;
+    try {
+      this.nodeDB = ReconSCMDBDefinition.NODES.getTable(scmDBStore);
+    } catch (IOException ex) {
+      LOG.debug("Failed to get NODES table. {}", ex.getMessage());
+    }
+  }
+
+  /**
+   * Flush the container replica history in-memory map to DB.
+   * @param clearMap true to clear the in-memory map after flushing completes.
+   */
+  public void flushReplicaHistoryMapToDB(boolean clearMap) {
+    if (replicaHistoryMap == null) {
+      return;
+    }
+    synchronized (replicaHistoryMap) {
+      try {
+        for (Map.Entry<Long, Map<UUID, ContainerReplicaHistory>> entry :
+            replicaHistoryMap.entrySet()) {
+          final long containerId = entry.getKey();
+          final Map<UUID, ContainerReplicaHistory> map = entry.getValue();
+          dbServiceProvider.storeContainerReplicaHistoryMap(containerId, map);
```

Comment:
Can we support a batch interface that uses the RocksDB BatchOperation? That will help with faster flushing.

Reply:
Good idea. I'll try